File: arm.64/langs/typescript/src/type.ts
import koffi from 'koffi';

export type Address = string | null;
export type Payload = string | null;
export type Options = string | null;

export type Sid = string;
export type Tag = string;

export type Fit = (address: Address, payload: Payload, options: Options) => boolean;
export type Fun = (address: Address, payload: Payload, options: Options) => number;

export interface Def {
  sid: Sid;
  tag: Tag;
  fit: unknown;
  fun: unknown;
}

export type MoreFn = (def: Def) => number;
export type PumpFn = (address: Address, payload: Payload, options: Options) => number;
export type LessFn = (def: Def) => number;

export interface Cabi {
  name: string;
  path: string;
  lib: ReturnType<typeof koffi.load>;
  More: MoreFn;
  Pump: PumpFn;
  Less: LessFn;
}

export interface Triplet {
  address: string;
  payload?: string;
  options?: string;
}

koffi.proto('bool FitFn(const char* address, const char* payload, const char* options)');
koffi.proto('int FunFn(const char* address, const char* payload, const char* options)');

export const DefStruct = koffi.struct('Def', {
  sid: 'const char*',
  tag: 'const char*',
  fit: 'FitFn*',
  fun: 'FunFn*',
});

export const DefPtr = koffi.pointer(DefStruct);

File: arm.64/langs/zig/src/type.zig
const std = @import("std");

pub const Address = ?[*:0]const u8;
pub const Payload = ?[*:0]const u8;
pub const Options = ?[*:0]const u8;
pub const Sid = ?[*:0]const u8;
pub const Tag = ?[*:0]const u8;

pub const FitFn = *const fn (address: Address, payload: Payload, options: Options) callconv(.c) bool;
pub const FunFn = *const fn (address: Address, payload: Payload, options: Options) callconv(.c) c_int;

pub const Def = extern struct {
    sid: Sid,
    tag: Tag,
    fit: ?FitFn,
    fun: ?FunFn,
};

pub const MoreFn = *const fn (def: *const Def) callconv(.c) c_int;
pub const PumpFn = *const fn (address: Address, payload: Payload, options: Options) callconv(.c) c_int;
pub const LessFn = *const fn (def: *const Def) callconv(.c) c_int;

pub const Cabi = struct {
    name: []const u8,
    path: []const u8,
    dyn_lib: std.DynLib,
    more: ?MoreFn,
    pump: PumpFn,
    less: ?LessFn,
};

pub const Triplet = struct {
    address: [:0]const u8,
    payload: [:0]const u8,
    options: [:0]const u8,
};
